#Analysis by Alan Tomusiak for RNA-Seq data generated by Yong-Ho in Q42021
#Primary comparisons are looking at SA-bGal (high) and SA-bGal (low) populations.

#Loads appropriate libraries.
library(Rsamtools)
library(Rsubread)
library(DESeq2)
library(apeglm)
library(magrittr)
library(M3C)
library(readr)
library(dplyr)
library(umap)
library(stringr)
library(biomaRt)
library(tidyr)
library(ggplot2)
library(ashr)
library(DEGreport)
library(clusterProfiler)
library(qvalue)
library("RColorBrewer")
library("pheatmap")
library(scales)
library(gplots)
library(devtools)
library(EnhancedVolcano)
library(msigdbr)
library(circlize)
library(ComplexHeatmap)

setwd("/home/atom/Desktop/Data/Yong-Ho/BAMs") #Sets directory.
BAMs <- list.files(pattern = "\\.BAM$") #Acquires all BAM files in directory.
#raw_counts <- featureCounts(BAMs,annot.inbuilt="hg38",isPairedEnd=TRUE,nthreads=5) #Count matrix generation.
#raw_count_matrix <- raw_counts$counts #Pulls matrix.
#write.csv(raw_count_matrix,"yongho_counts_raw.csv") #Writes count matrix for easier future loading.
#Read count matrix below.
raw_count_matrix <- read.csv("~/Desktop/Data/Yong-Ho/BAMs/yongho_counts_raw.csv", row.names=1) 
keep <- rowSums((raw_count_matrix)) >= 50 #Removes genes with low counts.
raw_count_matrix <- raw_count_matrix[keep,]
yongho_metadata <- read_csv("~/Desktop/Data/Yong-Ho/yongho_metadata.csv") #Pulls metadata.
yongho_metadata <- yongho_metadata %>% arrange(id) #Re-arranges by alphabetical order.
colnames(raw_count_matrix) <- yongho_metadata$id
yongho_metadata$age_sabgal <- paste(yongho_metadata$young_or_aged,yongho_metadata$sabgal_high_or_low)
yongho_metadata$young_or_aged <- as.factor(yongho_metadata$young_or_aged)
yongho_metadata$cm_or_te <- as.factor(yongho_metadata$cm_or_te)
yongho_metadata$sabgal_high_or_low <- as.factor(yongho_metadata$sabgal_high_or_low)
yongho_metadata$sex <- as.factor(yongho_metadata$sex)
yongho_metadata$cd8_or_cd4 <- as.factor(yongho_metadata$cd8_or_cd4)
yongho_metadata$age_sabgal <- as.factor(yongho_metadata$age_sabgal)
rownames(yongho_metadata) <- yongho_metadata$id

# The role of the below segment of code is to replace gene IDs with gene names in the count matrix.
gene_list <- rownames(raw_count_matrix) #Acquires full list of genes.
#httr::set_config(httr::config(ssl_verifypeer = FALSE)) #Networking work-around.
#Below: Acquires table of NCBI gene IDs and maps to HGNC gene symbols.
#id_to_name_mapping <- getBM(attributes = c("entrezgene_id","hgnc_symbol"), 
#                            filters = "entrezgene_id", values = gene_list,
#                            mart=useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl"))
#write.csv(id_to_name_mapping,"id_to_name_mapping.csv")
#Reads in mapping of NCBI gene IDs to HGNC gene symbols.
id_to_name_mapping <- read.csv("~/Desktop/Data/Yong-Ho/BAMs/id_to_name_mapping.csv", row.names=1)
matrix_ids <- rownames(raw_count_matrix)
matched_positions <- match(matrix_ids,id_to_name_mapping$entrezgene_id) #Finds valid matches in table.
matched_symbols <- id_to_name_mapping$hgnc_symbol[matched_positions]#Finds position of valid match.
unmatched_positions <- match(matched_symbols,NA) #Finds position of gene IDs with no matches gene name.
unmatched_positions_2 <- match(matched_symbols,"") #Finds position of unfound gene IDs (primarily loci)
unmatched_positions <- unmatched_positions | unmatched_positions_2 #Combines failed matches.
unmatched_positions <- !is.na(unmatched_positions ) #Maps failed matches from NAs to boolean.
matched_symbols[unmatched_positions] <- matrix_ids[unmatched_positions] #Maps in original gene ID for failures.
unique_symbols <- make.names(matched_symbols, unique = TRUE) #Creates unique identifier for each gene name.
rownames(raw_count_matrix) <- unique_symbols #Replaces IDs with names.

#Next few segments of code are purely to perform segmentation of the count matrix and to find
#top differentially expressed genes depending on category.
#Below are top DEGs for CD4/CD8.

good_donor_samples <- colnames(dds_norm)[!grepl("09",colnames(dds_norm),fixed=TRUE)]
raw_count_matrix <- raw_count_matrix[,good_donor_samples]
yongho_metadata <- yongho_metadata[good_donor_samples,]
rownames(yongho_metadata) <- yongho_metadata$id

dds <- DESeqDataSetFromMatrix(raw_count_matrix,
                                   colData = yongho_metadata,
                                   design = ~ age_sabgal) 
#Transforms count matrix into DESeq object.
dds <- DESeq(dds) #Performs DE analysis.
CD8_res <- results(dds, contrast=c("cd8_or_cd4","cd8","cd4")) 
CD8_resLFC <- lfcShrink(dds, res = CD8_res, contrast = "cd8_or_cd4",type="ashr")
plotMA(CD8_resLFC, ylim=c(-2,2))
CD8_resOrdered <- CD8_resLFC[order(CD8_resLFC$padj),]
top_hits_CD8 <- cd8_resOrdered[!is.na(CD8_resOrdered$padj) & CD8_resOrdered$padj < .05,]
write.csv(top_hits_CD8,"yongho_top_hits_cd8.csv")

#Below are top DEGs for SAbGal High/Low
sabgal_res <- results(dds, contrast=c("sabgal_high_or_low","high","low"))
sabgal_resLFC <- lfcShrink(dds, res = sabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(sabgal_resLFC, ylim=c(-2,2))
sabgal_resOrdered <- sabgal_resLFC[order(sabgal_resLFC$padj),]
top_hits_sabgal <- sabgal_resOrdered[!is.na(sabgal_resOrdered$padj) & sabgal_resOrdered$padj < .05,]
write.csv(top_hits_sabgal,"yongho_top_hits_sabgal.csv")

#Below are top DEGs for Old/Young
aged_res <- results(dds, contrast=c("young_or_aged","young","aged"))
aged_resLFC <- lfcShrink(dds, res = aged_res, contrast = "young_or_aged",type="ashr")
plotMA(aged_resLFC, ylim=c(-2,2))
aged_resOrdered <- aged_resLFC[order(aged_resLFC$padj),]
top_hits_aged <- aged_resOrdered[!is.na(aged_resOrdered$padj) & aged_resOrdered$padj < .05,]
write.csv(top_hits_aged,"top_hits_aged.csv")

#Next segment is to find high-level patterns in data via UMAP.
dds_norm <- vst(dds) #VST normalization.
normalized_counts <- assay(dds_norm) %>% t()
umap_results <- umap::umap(normalized_counts)
complete_umap <- data.frame(umap_results$layout) %>%
  tibble::rownames_to_column("id") %>%
  dplyr::inner_join(yongho_metadata, by = "id")
#CM or TE
ggplot(
  complete_umap,
  aes( x = X1, y = X2, color=cm_or_te)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction - Cell Type", x="Component 1", y="Component 2",
                                        color = "Central Memory or \n Terminal/Effector \n Memory") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))
#Donor ID
ggplot(
  complete_umap,
  aes( x = X1, y = X2, color=donor_id_long)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction - Donors", x="Component 1", y="Component 2",
                                        color = "Donor ID                   ") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))
#Age
ggplot(
  complete_umap,
  aes( x = X1, y = X2, color=young_or_aged)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction - Young vs. Older", x="Component 1", y="Component 2",
                                        color = "Donor Age                   ") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))
#SABGal High or Low
ggplot(
  complete_umap,
  aes( x = X1, y = X2, color=sabgal_high_or_low)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction - SABGal", x="Component 1", y="Component 2",
                                        color = "SABGal                   ") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))

#Below we segment into CM and TE, as those are the dominant clusters in the data.
#We will then proceed with looking at UMAP analysis for these separated populations.
cm_samples <- colnames(dds_norm)[grepl("CM",colnames(dds_norm),fixed=TRUE)]
te_samples <- colnames(dds_norm)[grepl("TE",colnames(dds_norm),fixed=TRUE)]
cm_counts <- assay(dds_norm)[,cm_samples ] %>% t()
te_counts <-  assay(dds_norm)[,te_samples] %>% t()

umap_results_cm <- umap::umap(cm_counts)
cm_umap <- data.frame(umap_results_cm$layout) %>%
  tibble::rownames_to_column("id") %>%
  dplyr::inner_join(yongho_metadata, by = "id")
ggplot(
  cm_umap,
  aes( x = X1, y = X2, color=young_or_aged)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction CM Only - SAbGal", x="Component 1", y="Component 2",
                                        color = "SAbGal High Or Low") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))

umap_results_te <- umap::umap(te_counts)
te_umap <- data.frame(umap_results_te$layout) %>%
  tibble::rownames_to_column("id") %>%
  dplyr::inner_join(yongho_metadata, by = "id")
ggplot(
  te_umap,
  aes( x = X1, y = X2, color=young_or_aged)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction EM Only - SAbGal", x="Component 1", y="Component 2",
                                        color = "SAbGal High Or Low") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))

#Below we segment into CD4 and CD8.
#We will then proceed with looking at UMAP analysis for these separated populations.
CD4_samples <- colnames(dds_norm)[grepl("CD4",colnames(dds_norm),fixed=TRUE)]
CD8_samples <- colnames(dds_norm)[grepl("CD8",colnames(dds_norm),fixed=TRUE)]
CD4_counts <- assay(dds_norm)[,CD4_samples ] %>% t()
CD8_counts <-  assay(dds_norm)[,CD8_samples] %>% t()

umap_results_CD4 <- umap::umap(CD4_counts)
CD4_umap <- data.frame(umap_results_CD4$layout) %>%
  tibble::rownames_to_column("id") %>%
  dplyr::inner_join(yongho_metadata, by = "id")
ggplot(
  CD4_umap,
  aes( x = X1, y = X2, color=sabgal_high_or_low)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction CD4 Only - SAbGal", x="Component 1", y="Component 2",
                                        color = "SAbGal High Or Low") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))

umap_results_CD8 <- umap::umap(CD8_counts)
CD8_umap <- data.frame(umap_results_CD8$layout) %>%
  tibble::rownames_to_column("id") %>%
  dplyr::inner_join(yongho_metadata, by = "id")
ggplot(
  CD8_umap,
  aes( x = X1, y = X2, color=sabgal_high_or_low)) +
  geom_point() + theme_classic() + labs(title="RNA-Seq UMAP Dimensional Reduction CD8 Only - SAbGal", x="Component 1", y="Component 2",
                                        color = "SAbGal High Or Low") +
  theme(legend.key.height= unit(1, 'cm'),
        legend.key.width= unit(1.5, 'cm'), legend.title = element_text(size=10))

#Below are top DEGs for CD4sabgal high/low
CD4_samples <- colnames(raw_count_matrix)[grepl("CD4",colnames(dds_norm),fixed=TRUE)]
CD4_counts <- raw_count_matrix[,CD4_samples ]
CD4_metadata <-  yongho_metadata[CD4_samples,]
CD4_dds <- DESeqDataSetFromMatrix(CD4_counts,
                                  colData = CD4_metadata,
                                  design = ~ age_sabgal)
CD4_dds <- DESeq(CD4_dds)
cd4sabgal_res <- results(CD4_dds, contrast=c("sabgal_high_or_low","high","low"))
cd4sabgal_resLFC <- lfcShrink(CD4_dds, res = cd4sabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(cd4sabgal_resLFC, ylim=c(-2,2))
cd4sabgal_resOrdered <- cd4sabgal_resLFC[order(cd4sabgal_resLFC$padj),]
top_hits_cd4sabgal <- cd4sabgal_resOrdered[!is.na(cd4sabgal_resOrdered$padj) & cd4sabgal_resOrdered$padj < .05,]

#Below are top DEGs for CD8sabgal high/low
CD8_samples <- colnames(raw_count_matrix)[grepl("CD8",colnames(dds_norm),fixed=TRUE)]
CD8_counts <- raw_count_matrix[,CD8_samples ]
CD8_metadata <-  yongho_metadata[CD8_samples,]
CD8_dds <- DESeqDataSetFromMatrix(CD8_counts,
                                  colData = CD8_metadata,
                                  design = ~ age_sabgal)
CD8_dds <- DESeq(CD8_dds)
cd8sabgal_res <- results(CD8_dds, contrast=c("sabgal_high_or_low","high","low"))
cd8sabgal_resLFC <- lfcShrink(CD8_dds, res = cd8sabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(cd8sabgal_resLFC, ylim=c(-2,2))
cd8sabgal_resOrdered <- cd8sabgal_resLFC[order(cd8sabgal_resLFC$padj),]
top_hits_cd8sabgal <- cd8sabgal_resOrdered[!is.na(cd8sabgal_resOrdered$padj) & cd8sabgal_resOrdered$padj < .05,]

#Below are top DEGs for CD4tesabgal high/low
CD4TE_samples <- colnames(raw_count_matrix)[grepl("CD4TE",colnames(dds_norm),fixed=TRUE)]
CD4TE_counts <- raw_count_matrix[,CD4TE_samples ]
CD4TE_metadata <-  yongho_metadata[CD4TE_samples,]
CD4TE_dds <- DESeqDataSetFromMatrix(CD4TE_counts,
                                    colData = CD4TE_metadata,
                                    design = ~ young_or_aged + sabgal_high_or_low +
                                      sex + age + pooled)
CD4TE_dds <- DESeq(CD4TE_dds)
cd4TEsabgal_res <- results(CD4TE_dds, contrast=c("sabgal_high_or_low","high","low"))
cd4TEsabgal_resLFC <- lfcShrink(CD4TE_dds, res = cd4TEsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(cd4TEsabgal_resLFC, ylim=c(-2,2))
cd4TEsabgal_resOrdered <- cd4TEsabgal_resLFC[order(cd4TEsabgal_resLFC$padj),]
top_hits_cd4TEsabgal <- cd4TEsabgal_resOrdered[!is.na(cd4TEsabgal_resOrdered$padj) & cd4TEsabgal_resOrdered$padj < .05,]
write.csv(top_hits_cd4TEsabgal,"yongho_top_hits_cd4TE_sabgal.csv")

#Below are top DEGs for CD4CMsabgal high/low
CD4CM_samples <- colnames(raw_count_matrix)[grepl("CD4CM",colnames(dds_norm),fixed=TRUE)]
CD4CM_counts <- raw_count_matrix[,CD4CM_samples ]
CD4CM_metadata <-  yongho_metadata[CD4CM_samples,]
CD4CM_dds <- DESeqDataSetFromMatrix(CD4CM_counts,
                                    colData = CD4CM_metadata,
                                    design = ~ age_sabgal)
CD4CM_dds <- DESeq(CD4CM_dds)
cd4CMsabgal_res <- results(CD4CM_dds, contrast=c("sabgal_high_or_low","high","low"))
cd4CMsabgal_resLFC <- lfcShrink(CD4CM_dds, res = cd4CMsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(cd4CMsabgal_resLFC, ylim=c(-2,2))
cd4CMsabgal_resOrdered <- cd4CMsabgal_resLFC[order(cd4CMsabgal_resLFC$padj),]
top_hits_cd4CMsabgal <- cd4CMsabgal_resOrdered[!is.na(cd4CMsabgal_resOrdered$padj) & cd4CMsabgal_resOrdered$padj < .05,]
write.csv(top_hits_cd4CMsabgal,"yongho_top_hits_cd4CM_sabgal.csv")

#Below are top DEGs for CD8TEsabgal high/low
CD8TE_samples <- colnames(raw_count_matrix)[grepl("CD8TE",colnames(dds_norm),fixed=TRUE)]
CD8TE_counts <- raw_count_matrix[,CD8TE_samples ]
CD8TE_metadata <-  yongho_metadata[CD8TE_samples,]
CD8TE_dds <- DESeqDataSetFromMatrix(CD8TE_counts,
                                    colData = CD8TE_metadata,
                                    design = ~ age_sabgal)
CD8TE_dds <- DESeq(CD8TE_dds)
cd8TEsabgal_res <- results(CD8TE_dds, contrast=c("sabgal_high_or_low","high","low"))
cd8TEsabgal_resLFC <- lfcShrink(CD8TE_dds, res = cd8TEsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(cd8TEsabgal_resLFC, ylim=c(-2,2))
cd8TEsabgal_resOrdered <- cd8TEsabgal_resLFC[order(cd8TEsabgal_resLFC$padj),]
top_hits_cd8TEsabgal <- cd8TEsabgal_resOrdered[!is.na(cd8TEsabgal_resOrdered$padj) & cd8TEsabgal_resOrdered$padj < .05,]
write.csv(top_hits_cd8TEsabgal,"yongho_top_hits_cd8TE_sabgal.csv")

#Below are top DEGs for CD8CMsabgal high/low
CD8CM_samples <- colnames(raw_count_matrix)[grepl("CD8CM",colnames(dds_norm),fixed=TRUE)]
CD8CM_counts <- raw_count_matrix[,CD8CM_samples ]
CD8CM_metadata <-  yongho_metadata[CD8CM_samples,]
CD8CM_dds <- DESeqDataSetFromMatrix(CD8CM_counts,
                                    colData = CD8CM_metadata,
                                    design = ~ young_or_aged + sabgal_high_or_low +
                                      sex + age + pooled)
CD8CM_dds <- DESeq(CD8CM_dds)
cd8CMsabgal_res <- results(CD8CM_dds, contrast=c("sabgal_high_or_low","high","low"))
cd8CMsabgal_resLFC <- lfcShrink(CD8CM_dds, res = cd8CMsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(cd8CMsabgal_resLFC, ylim=c(-2,2))
cd8CMsabgal_resOrdered <- cd8CMsabgal_resLFC[order(cd8CMsabgal_resLFC$padj),]
top_hits_cd8CMsabgal <- cd8CMsabgal_resOrdered[!is.na(cd8CMsabgal_resOrdered$padj) & cd8CMsabgal_resOrdered$padj < .05,]
write.csv(top_hits_cd8CMsabgal,"yongho_top_hits_cd8CM_sabgal.csv")

#Below are top DEGs for CD4TE aged vs. younger
cd4TEage_res <- results(CD4TE_dds, contrast=c("young_or_aged","aged","young"))
cd4TEage_resLFC <- lfcShrink(CD4TE_dds, res = cd4TEage_res, contrast = "young_or_aged",type="ashr")
plotMA(cd4TEage_resLFC, ylim=c(-2,2))
cd4TEage_resOrdered <- cd4TEage_resLFC[order(cd4TEage_resLFC$padj),]
top_hits_cd4TEage <- cd4TEage_resOrdered[!is.na(cd4TEage_resOrdered$padj) & cd4TEage_resOrdered$padj < .05,]
write.csv(top_hits_cd4TEage,"yongho_top_hits_cd4TEage.csv")

#Below are top DEGs for CD8TE aged vs. younger
cd8TEage_res <- results(CD8TE_dds, contrast=c("young_or_aged","aged","young"))
cd8TEage_resLFC <- lfcShrink(CD8TE_dds, res = cd8TEage_res, contrast = "young_or_aged",type="ashr")
plotMA(cd8TEage_resLFC, ylim=c(-2,2))
cd8TEage_resOrdered <- cd8TEage_resLFC[order(cd8TEage_resLFC$padj),]
top_hits_cd8TEage <- cd8TEage_resOrdered[!is.na(cd8TEage_resOrdered$padj) & cd8TEage_resOrdered$padj < .05,]
write.csv(top_hits_cd8TEage,"yongho_top_hits_cd8TEage.csv")

#Below are top DEGs for CD4CM aged vs. younger
cd4CMage_res <- results(CD4CM_dds, contrast=c("young_or_aged","aged","young"))
cd4CMage_resLFC <- lfcShrink(CD4CM_dds, res = cd4CMage_res, contrast = "young_or_aged",type="ashr")
plotMA(cd4CMage_resLFC, ylim=c(-2,2))
cd4CMage_resOrdered <- cd4CMage_resLFC[order(cd4CMage_resLFC$padj),]
top_hits_cd4CMage <- cd4CMage_resOrdered[!is.na(cd4CMage_resOrdered$padj) & cd4CMage_resOrdered$padj < .05,]
write.csv(top_hits_cd4CMage,"yongho_top_hits_cd4CMage.csv")

#Below are top DEGs for CD8CM aged vs. younger
cd8CMage_res <- results(CD8CM_dds, contrast=c("young_or_aged","aged","young"))
cd8CMage_resLFC <- lfcShrink(CD8CM_dds, res = cd8CMage_res, contrast = "young_or_aged",type="ashr")
plotMA(cd8CMage_resLFC, ylim=c(-2,2))
cd8CMage_resOrdered <- cd8CMage_resLFC[order(cd8CMage_resLFC$padj),]
top_hits_cd8CMage <- cd8CMage_resOrdered[!is.na(cd8CMage_resOrdered$padj) & cd8CMage_resOrdered$padj < .05,]
write.csv(top_hits_cd8CMage,"yongho_top_hits_cd8CMage.csv")

#Below are top DEGs old samples SABgal high vs. low
old_samples <- colnames(raw_count_matrix)[grepl("A",colnames(dds_norm),fixed=TRUE)]
old_counts <- raw_count_matrix[,old_samples ]
old_metadata <-  yongho_metadata[old_samples,]
old_dds <- DESeqDataSetFromMatrix(old_counts,
                                    colData = old_metadata,
                                    design = ~ sabgal_high_or_low + cm_or_te + cd8_or_cd4 +
                                      sex + pooled)
old_dds <- DESeq(old_dds)
oldsabgal_res <- results(old_dds, contrast=c("sabgal_high_or_low","high","low"))
oldsabgal_resLFC <- lfcShrink(old_dds, res = oldsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(oldsabgal_resLFC, ylim=c(-2,2))
oldsabgal_resOrdered <- oldsabgal_resLFC[order(oldsabgal_resLFC$padj),]
top_hits_oldsabgal <- oldsabgal_resOrdered[!is.na(oldsabgal_resOrdered$padj) & oldsabgal_resOrdered$padj < .05,]
write.csv(top_hits_oldsabgal,"yongho_top_hits_oldsabgal.csv")

#Below are top DEGs young samples SABgal high vs. low
young_samples <- colnames(raw_count_matrix)[grepl("Y",colnames(dds_norm),fixed=TRUE)]
young_counts <- raw_count_matrix[,young_samples ]
young_metadata <-  yongho_metadata[young_samples,]
young_dds <- DESeqDataSetFromMatrix(young_counts,
                                  colData = young_metadata,
                                  design = ~ sabgal_high_or_low + cm_or_te + cd8_or_cd4 +
                                     pooled)
young_dds <- DESeq(young_dds)
youngsabgal_res <- results(young_dds, contrast=c("sabgal_high_or_low","high","low"))
youngsabgal_resLFC <- lfcShrink(young_dds, res = youngsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(youngsabgal_resLFC, ylim=c(-2,2))
youngsabgal_resOrdered <- youngsabgal_resLFC[order(youngsabgal_resLFC$padj),]
top_hits_youngsabgal <- youngsabgal_resOrdered[!is.na(youngsabgal_resOrdered$padj) & youngsabgal_resOrdered$padj < .05,]
write.csv(top_hits_youngsabgal,"yongho_top_hits_youngsabgal.csv")

#Below are top DEGs for CD4tesabgal high/low older only
CD4TEold_samples <- colnames(old_counts)[grepl("CD4TE",colnames(old_dds),fixed=TRUE)]
CD4TEold_counts <- raw_count_matrix[,CD4TEold_samples]
CD4TEold_metadata <-  yongho_metadata[CD4TEold_samples,]
CD4TEold_dds <- DESeqDataSetFromMatrix(CD4TEold_counts,
                                    colData = CD4TEold_metadata,
                                    design = ~ sabgal_high_or_low +
                                      sex +  pooled)
CD4TEold_dds <- DESeq(CD4TEold_dds)
CD4TEoldsabgal_res <- results(CD4TEold_dds, contrast=c("sabgal_high_or_low","high","low"))
CD4TEoldsabgal_resLFC <- lfcShrink(CD4TEold_dds, res = CD4TEoldsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(CD4TEoldsabgal_resLFC, ylim=c(-2,2))
CD4TEoldsabgal_resOrdered <- CD4TEoldsabgal_resLFC[order(CD4TEoldsabgal_resLFC$padj),]
top_hits_CD4TEoldsabgal <- CD4TEoldsabgal_resOrdered[!is.na(CD4TEoldsabgal_resOrdered$padj) & CD4TEoldsabgal_resOrdered$padj < .05,]
write.csv(top_hits_CD4TEoldsabgal,"yongho_top_hits_CD4TEold_sabgal.csv")

#Below are top DEGs for CD4tesabgal high/low younger only
CD4TEyoung_samples <- colnames(young_counts)[grepl("CD4TE",colnames(young_dds),fixed=TRUE)]
CD4TEyoung_counts <- raw_count_matrix[,CD4TEyoung_samples]
CD4TEyoung_metadata <-  yongho_metadata[CD4TEyoung_samples,]
CD4TEyoung_dds <- DESeqDataSetFromMatrix(CD4TEyoung_counts,
                                       colData = CD4TEyoung_metadata,
                                       design = ~ sabgal_high_or_low +
                                         pooled)
CD4TEyoung_dds <- DESeq(CD4TEyoung_dds)
CD4TEyoungsabgal_res <- results(CD4TEyoung_dds, contrast=c("sabgal_high_or_low","high","low"))
CD4TEyoungsabgal_resLFC <- lfcShrink(CD4TEyoung_dds, res = CD4TEyoungsabgal_res, contrast = "sabgal_high_or_low",type="ashr")
plotMA(CD4TEyoungsabgal_resLFC, ylim=c(-2,2))
CD4TEyoungsabgal_resOrdered <- CD4TEyoungsabgal_resLFC[order(CD4TEyoungsabgal_resLFC$padj),]
top_hits_CD4TEyoungsabgal <- CD4TEyoungsabgal_resOrdered[!is.na(CD4TEyoungsabgal_resOrdered$padj) & CD4TEyoungsabgal_resOrdered$padj < .05,]
write.csv(top_hits_CD4TEyoungsabgal,"yongho_top_hits_CD4TEyoung_sabgal.csv")

#Looking at boxplots for interesting genes with age (subsetted by CM or TE)
interesting_gene_with_age <- plotCounts(dds, gene="B3GAT1", intgroup=c("cm_or_te","young_or_aged"),returnData=TRUE)
ggplot(interesting_gene_with_age, aes(x=young_or_aged, y=count,fill=cm_or_te)) + 
  geom_boxplot() + theme_classic() + labs(title="CD57 Expression On T Cell Subsets in Older or Younger Donors",x = "Age Group",
  y="Counts",fill="Central Memory (CM) \n or Terminal + \n Effector Memory (TE)")

#Looking at boxplots for interesting genes with SABGAL (subsetted by CM or TE)
interesting_gene_with_sabgal <- plotCounts(dds, gene="B3GAT1", intgroup=c("cm_or_te","sabgal_high_or_low"),returnData=TRUE)
ggplot(interesting_gene_with_sabgal, aes(x=sabgal_high_or_low, y=count,fill=cm_or_te)) + 
  geom_boxplot() + theme_classic() + labs(title="CD57 Expression On T Cell Subsets in SA-bGal High Or Low Cells",x = "SAbGal High or Low",
                                          y="Counts",fill="Central Memory (CM) \n or Terminal + \n Effector Memory (TE)")

#Looking at boxplots for interesting genes with age (subsetted by CM or TE) - CD8s only
interesting_gene_with_age_cd8 <- plotCounts(CD8_dds, gene="CXCL10", intgroup=c("cm_or_te","young_or_aged"),returnData=TRUE)
ggplot(interesting_gene_with_age_cd8, aes(x=young_or_aged, y=count,fill=cm_or_te)) + 
  geom_boxplot() + theme_classic() + labs(title="CXCL10 Expression On T Cell Subsets in Older or Younger Donors - CD8s only",x = "Age Group",
                                          y="Counts",fill="Central Memory (CM) \n or Terminal + \n Effector Memory (TE)")

#Looking at boxplots for interesting genes with SABGAL (subsetted by CM or TE) - CD8s only
interesting_gene_with_sabgal_cd8 <- plotCounts(CD8_dds, gene="CXCL10", intgroup=c("cm_or_te","sabgal_high_or_low"),returnData=TRUE)
ggplot(interesting_gene_with_sabgal_cd8, aes(x=sabgal_high_or_low, y=count,fill=cm_or_te)) + 
  geom_boxplot() + theme_classic() + labs(title="CXCL10 Expression On T Cell Subsets in SA-bGal High Or Low Cells - CD8s only",x = "SAbGal High or Low",
                                          y="Counts",fill="Central Memory (CM) \n or Terminal + \n Effector Memory (TE)")

#Looking at boxplots for interesting genes with age (subsetted by CM or TE) - CD4s only
interesting_gene_with_age_cd4 <- plotCounts(CD4_dds, gene="CXCL10", intgroup=c("cm_or_te","young_or_aged"),returnData=TRUE)
ggplot(interesting_gene_with_age_cd4, aes(x=young_or_aged, y=count,fill=cm_or_te)) + 
  geom_boxplot() + theme_classic() + labs(title="CXCL10 Expression On T Cell Subsets in Older or Younger Donors - CD4s only",x = "Age Group",
                                          y="Counts",fill="Central Memory (CM) \n or Terminal + \n Effector Memory (TE)")

#Looking at boxplots for interesting genes with SABGAL (subsetted by CM or TE) - CD4s only
interesting_gene_with_sabgal_cd4 <- plotCounts(CD4_dds, gene="CXCL10", intgroup=c("cm_or_te","sabgal_high_or_low"),returnData=TRUE)
ggplot(interesting_gene_with_sabgal_cd4, aes(x=sabgal_high_or_low, y=count,fill=cm_or_te)) + 
  geom_boxplot() + theme_classic() + labs(title="CXCL10 Expression On T Cell Subsets in SA-bGal High Or Low Cells - CD4s only",x = "SAbGal High or Low",
                                          y="Counts",fill="Central Memory (CM) \n or Terminal + \n Effector Memory (TE)")

#Summarization function cribbed from Goldfar Lab
summarySE <- function(data=NULL, measurevar, groupvars=NULL, na.rm=FALSE,
                      conf.interval=.95, .drop=TRUE, removeOutliers=F, logt=F) {
  length2 <- function (x, na.rm=FALSE) {
    if (na.rm) sum(!is.na(x))
    else       length(x)
  }
  datac <- plyr::ddply(data, groupvars, .drop=.drop,
                       .fun = function(xx, col) {
                         
                         if (removeOutliers) {
                           print('pre')
                           print(xx[[col]])
                           #xx[[col]] <- boxB(xx[[col]])
                           print('post')
                           #print(exp(rm.outlier(log(xx[[col]]), fill = FALSE, median = FALSE, opposite = FALSE)))
                           print(scores(log(xx[[col]]), type="t", prob=0.90) )
                           #print(xx[[col]][boxB(xx[[col]], method="resistant", logt=logt, k=2.5)[["outliers"]]])
                         }
                         c(N    = length2(xx[[col]], na.rm=na.rm),
                           mean = mean   (xx[[col]], na.rm=na.rm),
                           sd   = sd     (xx[[col]], na.rm=na.rm)
                         )
                       },
                       measurevar
  )
  datac <- plyr::rename(datac, c("mean" = measurevar))
  datac$se <- datac$sd / sqrt(datac$N)  # Calculate standard error of the mean
  ciMult <- qt(conf.interval/2 + .5, datac$N-1)
  datac$ci <- datac$se * ciMult
  return(datac)
}

#Quick visualization and summarization..
count_data <- trial %>% t()
data_complete <- data.frame(count_data ) %>%
  tibble::rownames_to_column("id") %>%
  dplyr::inner_join(yongho_metadata, by = "id")
data_summary <- summarySE(data_complete,"KLRG1",c("young_or_aged","sabgal_high_or_low"))
CD8CM_counts["KLRG1",]
ggplot(data_summary, aes(x=young_or_aged, y=KLRG1, fill=sabgal_high_or_low)) +
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=KLRG1-se,ymax=KLRG1+se),position = position_dodge(width = .9),width=.2) + 
  theme_classic() + 
  labs(title="KLRG1 Expression On T Cell Subsets in SA-bGal High Or Low CD8+ CM Cells",x = "Younger or Older",
                                          y="Normalized Counts",
       fill="SAbGal High or Low") +
  scale_fill_manual(values=c("pink","firebrick4"))
  
#Let's use some of Marius's code for cool heatmaps..
sabgal_resLFC$qvalue<- qvalue(sabgal_resLFC$pvalue)$qvalue
cd8_resLFC$qvalue <- qvalue(cd8_resLFC$pvalue)$qvalue
res_Infection_in_WT$qvalue<- qvalue(res_Infection_in_WT$pvalue)$qvalue
res_Infection_in_KO$qvalue<- qvalue(res_Infection_in_KO$pvalue)$qvalue
plotDispEsts(dds, main="Dispersion plot")
trial <- apply(CD8CM_counts, 1, rescale) %>% t()
interesting_genes <- c("RGS3",
                       "KLRD1",
                       "FCGR3A",
                       "CXCR5",
                       "HLA.DRB5",
                       "HLA.DQA2",
                       "LYN",
                       "GAS7", "FOXP3", "IL2RA", "IL7R")
df <- as.data.frame(colData(dds)[,c("sabgal_high_or_low","young_or_aged","donor_id_long")])
pheatmap(trial[interesting_genes,], cluster_rows=TRUE, show_rownames=TRUE, show_colnames=FALSE,
         cluster_cols=TRUE, annotation_col=df,labs_row="PDCD1")
sampleDists <- dist(t(assay(dds_norm)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(dds_norm$condition, dds_norm$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
colnames(sampleDistMatrix) <- rownames(yongho_metadata)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

dds_lrt <- DESeq(CD4CM_dds, test="LRT", reduced = ~ 1)
rld <- vst(dds_lrt)
res_LRT <- results(dds_lrt)
res_LRT$qvalue <- qvalue(res_LRT$pvalue)$qvalue
top_hits_overall <- res_LRT[!is.na(res_LRT$qvalue) & res_LRT$qvalue < .1
                         #   & ((res_LRT$log2FoldChange > 1  | res_LRT$log2FoldChange < -1))
                            ,]
top_hits_ordered <- rownames(top_hits_overall[order(top_hits_overall$qvalue),])
rld_mat <- assay(rld)
Selected_rld <- rld_mat[top_hits_ordered,]
rownames(CD4CM_metadata) <- CD4CM_metadata$id
clusters <- degPatterns(Selected_rld, CD4CM_metadata, time ='age_sabgal',col='young_or_aged')
p <- clusters$plot +
  theme_classic() +
  labs(x="SAbGal High or Low", title="Cluster analysis of DE Genes - CD4 CM+ Only") +
  scale_x_discrete("SAbGal High or Low", labels = c("aged high" = "High","aged low" = "Low", 
                                                  "young high" = "High","young low" = "Low")) +
  scale_fill_manual(values=c("pink","firebrick4"))
p

assayCluster <- rld_mat[clusters$df$genes,]
my_palette <- colorRampPalette(c("blue",'white','red'))(n=1000)
assayCluster <- assayCluster[,c(1,3,2,4,5,7,9,11,6,8,10,12)]
heatmap.2(assayCluster, col=my_palette,
          scale="row", key=F, keysize=1, symkey=T,
          density.info="none", trace="none",
          cexCol=0.6, labRow=F, dendrogram="row",Rowv=TRUE,Colv=FALSE,
          main="CD4+ CM q < .05 and log2fc > 1 or < -1 ")

dds_lrt <- DESeq(CD8TE_dds, test="LRT", reduced = ~ 1)
rld <- vst(dds_lrt)
res_LRT <- results(dds_lrt)
res_LRT$qvalue <- qvalue(res_LRT$pvalue)$qvalue
top_hits_overall <- res_LRT[!is.na(res_LRT$qvalue) & res_LRT$qvalue < .05 
                            & ((res_LRT$log2FoldChange > 1  | res_LRT$log2FoldChange < -1))
                            ,]
top_hits_ordered <- rownames(top_hits_overall[order(top_hits_overall$qvalue),])
rld_mat <- assay(rld)
Selected_rld <- rld_mat[top_hits_ordered,]
rownames(CD8TE_metadata) <- CD8TE_metadata$id
clusters <- degPatterns(Selected_rld, CD8TE_metadata, time ='age_sabgal',col='young_or_aged')
p <- clusters$plot +
  theme_classic() +
  labs(x="SAbGal High or Low", title="Cluster analysis of DE Genes - CD8+ TE Only") +
  scale_x_discrete("SAbGal High or Low", labels = c("aged high" = "High","aged low" = "Low", 
                                                    "young high" = "High","young low" = "Low")) +
  scale_fill_manual(values=c("pink","firebrick4"))
p

assayCluster <- rld_mat[clusters$df$genes,]
my_palette <- colorRampPalette(c("blue",'white','red'))(n=1000)
assayCluster <- assayCluster[,c(1,3,2,4,5,7,9,11,6,8,10,12)]
heatmap.2(assayCluster, col=my_palette,
          scale="row", key=F, keysize=1, symkey=T,
          density.info="none", trace="none",
          cexCol=0.6, labRow=F, dendrogram="row",Rowv=TRUE,Colv=FALSE,
          main="CD8+ TE q < .05 and log2fc > 1 or < -1 ")

dds_lrt <- DESeq(CD8_dds, test="LRT", reduced = ~ 1)
rld <- vst(dds_lrt)
res_LRT <- results(dds_lrt)
res_LRT$qvalue <- qvalue(res_LRT$pvalue)$qvalue
top_hits_overall <- res_LRT[!is.na(res_LRT$qvalue) & res_LRT$qvalue < .05 
                            & ((res_LRT$log2FoldChange > 1  | res_LRT$log2FoldChange < -1))
                            ,]
top_hits_ordered <- rownames(top_hits_overall[order(top_hits_overall$qvalue),])
rld_mat <- assay(rld)
Selected_rld <- rld_mat[top_hits_ordered,]
rownames(CD8_metadata) <- CD8_metadata$id
clusters <- degPatterns(Selected_rld, CD8_metadata, time ='age_sabgal',col='young_or_aged')
p <- clusters$plot +
  theme_classic() +
  labs(x="SAbGal High or Low", title="Cluster analysis of DE Genes - CD8+ Only") +
  scale_x_discrete("SAbGal High or Low", labels = c("aged high" = "High","aged low" = "Low", 
                                                    "young high" = "High","young low" = "Low")) +
  scale_fill_manual(values=c("pink","firebrick4"))
p

assayCluster <- rld_mat[clusters$df$genes,]
my_palette <- colorRampPalette(c("blue",'white','red'))(n=1000)
assayCluster <- assayCluster[,c(1,3,5,7,2,4,6,8,11,13,15,17,19,10,12,14,16,18,20)]
heatmap.2(assayCluster, col=my_palette,
          scale="row", key=F, keysize=1, symkey=T,
          density.info="none", trace="none",
          cexCol=0.6, labRow=F, dendrogram="row",Rowv=TRUE,Colv=FALSE,
          main="CD8+ q < .05 and log2fc > 1 or < -1 ")

dds_lrt <- DESeq(CD4_dds, test="LRT", reduced = ~ 1)
rld <- vst(dds_lrt)
res_LRT <- results(dds_lrt)
res_LRT$qvalue <- qvalue(res_LRT$pvalue)$qvalue
top_hits_overall <- res_LRT[!is.na(res_LRT$qvalue) & res_LRT$qvalue < .1 
                     #       & ((res_LRT$log2FoldChange > 1  | res_LRT$log2FoldChange < -1))
                            ,]
top_hits_ordered <- rownames(top_hits_overall[order(top_hits_overall$qvalue),])
rld_mat <- assay(rld)
Selected_rld <- rld_mat[top_hits_ordered,]
rownames(CD4_metadata) <- CD4_metadata$id
clusters <- degPatterns(Selected_rld, CD4_metadata, time ='age_sabgal',col='young_or_aged')
p <- clusters$plot +
  theme_classic() +
  labs(x="SAbGal High or Low", title="Cluster analysis of DE Genes - CD4+ Only") +
  scale_x_discrete("SAbGal High or Low", labels = c("aged high" = "High","aged low" = "Low", 
                                                    "young high" = "High","young low" = "Low")) +
  scale_fill_manual(values=c("pink","firebrick4"))
p

assayCluster <- rld_mat[clusters$df$genes,]
my_palette <- colorRampPalette(c("blue",'white','red'))(n=1000)
assayCluster <- assayCluster[,c(1,3,5,7,2,4,6,8,11,13,15,17,19,10,12,14,16,18,20)]
heatmap.2(assayCluster, col=my_palette,
          scale="row", key=F, keysize=1, symkey=T,
          density.info="none", trace="none",
          cexCol=0.6, labRow=F, dendrogram="row",Rowv=TRUE,Colv=FALSE,
          main="CD4+ q < .05 and log2fc > 1 or < -1 ")


dds_lrt <- DESeq(dds, test="LRT", reduced = ~ 1)
rld <- vst(dds_lrt)
res_LRT <- results(dds_lrt)
res_LRT$qvalue <- qvalue(res_LRT$pvalue)$qvalue
top_hits_overall <- res_LRT[!is.na(res_LRT$qvalue) & res_LRT$padj < .1
                            & ((res_LRT$log2FoldChange > .1  | res_LRT$log2FoldChange < -.1))
                            ,]
top_hits_ordered <- rownames(top_hits_overall[order(top_hits_overall$qvalue),])
rld_mat <- assay(rld)
Selected_rld <- rld_mat[top_hits_ordered,]
rownames(yongho_metadata) <- yongho_metadata$id
clusters <- degPatterns(Selected_rld, yongho_metadata, time ='age_sabgal',col='young_or_aged')
p <- clusters$plot +
  theme_classic() +
  labs(x="SAbGal High or Low", title="Cluster analysis of DE Genes") +
  scale_x_discrete("SAbGal High or Low", labels = c("aged high" = "High","aged low" = "Low", 
                                                    "young high" = "High","young low" = "Low")) +
  scale_fill_manual(values=c("pink","firebrick4"))
p

assayCluster <- rld_mat[clusters$df$genes,]
my_palette <- colorRampPalette(c("blue",'white','red'))(n=1000)
assayCluster <- assayCluster[,c(1,3,5,7,9,11,13,15,2,4,6,8,10,12,14,16,17,19,21,23,25,27,29,31,33,35,37,39,41,18,20,22,
                                24,26,28,30,32,34,36,38,40,42)]
heatmap.2(assayCluster, col=my_palette,
          scale="row", key=F, keysize=1, symkey=T,
          density.info="none", trace="none",
          cexCol=0.6, labRow=F, dendrogram="row",Rowv=TRUE,Colv=FALSE,
          main="q < .05 and log2fc > 1 or < -1 ")

msig <- msigdbr(species = "Homo sapiens") %>% filter(gs_cat == "H" | (gs_cat == "C2" & gs_subcat == "CP:REACTOME"))   %>% select(gs_name,human_gene_symbol)

#Find gene sets enriched in cluster 1 to 8
enrichment.results <- tibble()
for (clusterID in c(1, 2, 3,4,5,6)) {
  geneIds.cluster <-  clusters$df[(clusters$df$cluster ==clusterID),]
  em <- enricher(gene=geneIds.cluster$genes,TERM2GENE=msig, minGSSize = 8, qvalueCutoff=0.05)
  #write.table(clusters$df$symbol[(clusters$df$cluster ==clusterID)],sep=" ",row.names=F,col.names=F ,file=paste("cluster_",clusterID,".txt"))
  
  if (!is.null(em)) {
    result <- em@result
    result$cluster <- clusterID
    result <- filter(result, qvalue <= 0.05)
    if (nrow(result) > 0) {
      result$genes <- apply(result, 1, function(x) {str_c(sort(unlist(str_split(x[["geneID"]],"/"))),collapse=",")} )
      enrichment.results <- rbind(enrichment.results, result)
    }
  }
}
enrichment.results$category <- ""
enrichment.results$category[str_detect(enrichment.results$Description, "^HALLMARK")] <- "Hallmark"
enrichment.results$category[str_detect(enrichment.results$Description, "^REACTOME")] <- "Reactome"
enrichment.results$category[str_detect(enrichment.results$Description, "^GO")] <- "GO_CC"
table(enrichment.results$category)

#Exemple to get enriched gene sets and genes in cluster 8
enrichment.results$Description[enrichment.results$cluster == 4]
enrichment.results$genes[enrichment.results$cluster == 3]

num.in.cluster <- data.frame(table(clusters$df$cluster))
names(num.in.cluster) <- c("cluster", "n")
enrichment.results$ratio <- enrichment.results$Count / num.in.cluster$n[enrichment.results$cluster]

# convert to matrix format
enrichment.results_wide = pivot_wider(enrichment.results, id_cols = 
                                        c(Description, category), names_from = cluster, 
                                      values_from=c(pvalue, p.adjust, qvalue, Count, 
                                                    GeneRatio, BgRatio, ratio, genes))
include_sets <- c(
  #"REACTOME_INTERFERON_ALPHA_BETA_SIGNALING", 
  "REACTOME_IMMUNOREGULATORY_INTERACTIONS_BETWEEN_A_LYMPHOID_AND_A_NON_LYMPHOID_CELL",
  "REACTOME_METABOLISM_OF_AMINO_ACIDS_AND_DERIVATIVES",
  "HALLMARK_IL6_JAK_STAT3_SIGNALING",
  "REACTOME_CELL_CYCLE_CHECKPOINTS",
  "REACTOME_M_PHASE",
  "REACTOME_TRANSLATION",
  "REACTOME_NONSENSE_MEDIATED_DECAY_NMD", 
  "REACTOME_CELL_CYCLE_MITOTIC",  
  "REACTOME_CHEMOKINE_RECEPTORS_BIND_CHEMOKINES",
  "REACTOME_NEF_AND_SIGNAL_TRANSDUCTION" ,
  #"REACTOME_INTERFERON_ALPHA_BETA_SIGNALING", 
  "REACTOME_NEUTROPHIL_DEGRANULATION",
  "HALLMARK_PROTEIN_SECRETION",
  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  "REACTOME_ANTIGEN_PRESENTATION_FOLDING_ASSEMBLY_AND_PEPTIDE_LOADING_OF_CLASS_I_MHC",
  "REACTOME_ER_QUALITY_CONTROL_COMPARTMENT_ERQC",
  "REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION",
  "REACTOME_CHOLESTEROL_BIOSYNTHESIS",
  "REACTOME_GLUCOSE_METABOLISM",
  "HALLMARK_INFLAMMATORY_RESPONSE",
  "HALLMARK_INTERFERON_ALPHA_RESPONSE",
  "REACTOME_CELLULAR_RESPONSE_TO_HEAT_STRESS",
  "HALLMARK_CHOLESTEROL_HOMEOSTASIS",
  "HALLMARK_PROTEIN_SECRETION",
  "HALLMARK_MTORC1_SIGNALING" ,
  "HALLMARK_HYPOXIA" ,
  "HALLMARK_INTERFERON_GAMMA_RESPONSE" ,
  #"HALLMARK_ESTROGEN_RESPONSE_EARLY",
  #"HALLMARK_ESTROGEN_RESPONSE_LATE",
  "HALLMARK_FATTY_ACID_METABOLISM",
  "HALLMARK_P53_PATHWAY",
  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  "HALLMARK_COAGULATION",
  "HALLMARK_OXIDATIVE_PHOSPHORYLATION" 
)

enrichment.results.pruned <- filter(enrichment.results_wide, Description %in% c(include_sets))
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_CELL_CYCLE_CHECKPOINTS")] <- "Cell cycle checkpoints"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_METABOLISM_OF_AMINO_ACIDS_AND_DERIVATIVES")] <- "Metabolism of Amino Acids"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_IL6_JAK_STAT3_SIGNALING")] <- "IL6 JAK-STAT Signaling"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_CELL_CYCLE_CHECKPOINTS")] <- "Cell Cycle Checkpoints"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_M_PHASE")] <- "Mitosis"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_TRANSLATION")] <- "Translation"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_NONSENSE_MEDIATED_DECAY_NMD")] <- "Nonsense Mediated Decay"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_CHEMOKINE_RECEPTORS_BIND_CHEMOKINES")] <- "Chemokine Receptor Binding"    
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_NEF_AND_SIGNAL_TRANSDUCTION")] <- "NEF and Signal Transduction"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_NEUTROPHIL_DEGRANULATION")] <- "Neutrophil Degranulation"    
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_PROTEIN_SECRETION")] <- "Protein Secretion"    
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_TNFA_SIGNALING_VIA_NFKB")] <- "TNFa Signaling via NFkB"    
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_ANTIGEN_PRESENTATION_FOLDING_ASSEMBLY_AND_PEPTIDE_LOADING_OF_CLASS_I_MHC")] <- "Antigen presentation by MHC class I"    
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_ER_QUALITY_CONTROL_COMPARTMENT_ERQC")] <- "ER quality control compartment"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION")] <- "Extracellular matrix organisation"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_CHOLESTEROL_BIOSYNTHESIS")] <- "Cholesterol Biosynthesis"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_GLUCOSE_METABOLISM")] <- "Glucose Metabolism"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_INFLAMMATORY_RESPONSE")] <- "Inflammatory Response"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_INTERFERON_ALPHA_RESPONSE")] <- "Interferon Alpha and Beta response"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_CELLULAR_RESPONSE_TO_HEAT_STRESS")] <- "Cellular Response to Heat Stress"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_CHOLESTEROL_HOMEOSTASIS")] <- "Cholesterol Homeostasis"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_PROTEIN_SECRETION")] <- "Protein Secretion"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_MTORC1_SIGNALING")] <- "mTORC1 Signaling"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_HYPOXIA")] <- "Hypoxia"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_INTERFERON_GAMMA_RESPONSE")] <- "Interferon Gamma response"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_FATTY_ACID_METABOLISM")] <- "Fatty Acid Metabolism"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_P53_PATHWAY")] <- "P53 Pathway"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_COAGULATION")] <- "Coagulation"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "HALLMARK_OXIDATIVE_PHOSPHORYLATION")] <- "Oxidative Phosphorilation"
enrichment.results.pruned$Description[which(enrichment.results.pruned$Description == "REACTOME_IMMUNOREGULATORY_INTERACTIONS_BETWEEN_A_LYMPHOID_AND_A_NON_LYMPHOID_CELL")] <- "Interactions between a Lymphoid and a Non-Lymphoid Cell"

enrichment.results.pruned <- arrange(enrichment.results.pruned, category)
em.q.values <- select(enrichment.results.pruned, grep("qvalue", colnames(enrichment.results.pruned), value=T))

em.q.values <- em.q.values[, c("qvalue_1", "qvalue_2", "qvalue_3", "qvalue_4", "qvalue_5","qvalue_6")]
rownames(em.q.values) <- as.character(enrichment.results.pruned$Description)
colnames(em.q.values) <- c(1,2,3,4,5,6)
em.q.values <- -log10(as.matrix(em.q.values))

ratio.values <- select(enrichment.results.pruned, grep("ratio", colnames(enrichment.results.pruned), value=T))
ratio.values <- ratio.values[, c("ratio_1", "ratio_2", "ratio_3", "ratio_4", "ratio_5", "ratio_6")]

colnames(ratio.values) <- c(1,2,3,4,5,6)
ratio.values <- data.frame(ratio.values)

num.clusters <-6
max.q = 5 #max(em.q.values, na.rm=T)
min.q = min(em.q.values, na.rm=T)
#col_fun = colorRamp2(c(min.q, max.q, max.q), c("#300101", "#e41a1c", "#e41a1c"))
col_fun = colorRamp2(c(0, 2, 10), c("white",  "#e41a1c", "black"))

col.annot <- HeatmapAnnotation(cluster = anno_simple(colnames(em.q.values),
                                                     pt_gp = gpar(fontsize = 6),
                                                     height = unit(1, "mm"),
                                                     col = structure(brewer.pal(num.clusters, "Set3"), 
                                                                     names = colnames(em.q.values))),
                               show_annotation_name = F,
                               show_legend = F,
                               annotation_name_gp = gpar(fontsize = 6))


h <- Heatmap(em.q.values, 
             # labels
             column_title = "Enriched gene sets",
             column_title_gp = gpar(fontsize=7),
             row_names_gp = gpar(fontsize = 7),
             column_names_gp = gpar(fontsize = 6),
             column_names_rot = 0,
             row_title_gp = gpar(fontsize=7),
             #name = "-log10(q-value)",
             # legends
             bottom_annotation = col.annot,
             show_heatmap_legend = T,
             col = col_fun,
             heatmap_legend_param = list(color_bar = "continuous",
                                         title_gp = gpar(fontsize = 6),
                                         labels_gp = gpar(fontsize = 6),
                                         grid_width = unit(2,"mm")),
             
             cluster_rows = F,
             cluster_columns = F,
             show_row_names = T,
             show_column_names = T,
             #split = enrichment.results.pruned$category,
             cell_fun = function(j, i, x, y, width, height, fill) {
               grid.rect(x = x, y = y, width = width, height = height, 
                         gp = gpar(fill = "white", col = "#EEEEEE"))
               grid.circle(x = x, y = y, r = sqrt(ratio.values[i, j]) * 0.1,
                           gp = gpar(fill = col_fun(em.q.values[i, j]), col = NA))
             }
)
h